migrate/src/main/scala/migrate/internal/MigrationFile.scala

```diff
@@ -0,0 +1,62 @@
+package migrate.internal
+
+import interfaces.CompilationUnit
+import scalafix.interfaces._
+import migrate.utils.FileUtils
+import scala.util.{Try, Failure, Success}
+import migrate.utils.ScalaExtensions._
+import migrate.AbsolutePath
+
+private[migrate] trait MigrationFile {
+  def source: AbsolutePath
+  def patches: Seq[ScalafixPatch]
+  def previewAllPatches(): Try[CompilationUnit]
+  def previewPatches(patches: Seq[ScalafixPatch]): Try[CompilationUnit]
+}
+
+private[migrate] object MigrationFile {
+  def apply(source: AbsolutePath, evaluation: Option[ScalafixFileEvaluation]): MigrationFile = {
+    evaluation
+      .map(ScalafixMigrationFile(source, _))
+      .getOrElse(DefaultMigrationFile(source))
+  }
+}
+
+private[migrate] case class ScalafixMigrationFile(source: AbsolutePath, evaluation: ScalafixFileEvaluation) extends MigrationFile {
+  val patches: Seq[ScalafixPatch] = evaluation.getPatches().toSeq
+
+  def previewAllPatches(): Try[CompilationUnit] = {
+    evaluation.previewPatches()
+      .map { content => 
+        new CompilationUnit(source.value, content)
```

Comment:
I would delete the trait FileMigrationResult and change MigrationFile to a sealed trait FileMigrationState (initial, succeeded, and failed)
the migrate method would be on the initial State.

It will remove some confusion there is on FileMigration vs MigrationFile
```
sealed trait FileMigrationState {
  val evaluation: ScalafixFileEvaluation

  def source: AbsolutePath = AbsolutePath.from(evaluation.getEvaluatedFile).get

  def patches: Seq[ScalafixPatch] = evaluation.getPatches().toSeq

  def previewAllPatches(): Try[CompilationUnit] = {
    evaluation.previewPatches()
      .map { content =>
        new CompilationUnit(source.value, content)
      }
      .asScala
      .toTry(new ScalafixException(s"Cannot apply patch on file $source"))
  }

  def previewPatches(patches: Seq[ScalafixPatch]): Try[CompilationUnit] = {
    evaluation.previewPatches(patches.toArray)
      .map { content =>
        new CompilationUnit(source.value, content)
      }
      .asScala
      .toTry(new ScalafixException(s"Cannot apply patch on file $source"))
  }
}


object FileMigrationState {
  case class Initial(evaluation: ScalafixFileEvaluation) extends FileMigrationState {
    def migrate(compiler: Scala3Compiler): FileMigrationState.FinalState = {
      new FileMigration(this, compiler).migrate()
    }

    def success(necessaryPatches: Seq[ScalafixPatch]): Succeeded = Succeeded(evaluation, necessaryPatches)

    def failed(cause: Throwable): Failed = Failed(evaluation, cause)
  }
  sealed trait FinalState
  case class Succeeded(evaluation: ScalafixFileEvaluation, necessaryPatched: Seq[ScalafixPatch]) extends FileMigrationState with FinalState

  case class Failed(evaluation: ScalafixFileEvaluation, cause: Throwable) extends FileMigrationState with FinalState
}
```

Reply:
I like the idea.